File: Linked List/21. Merge Two Sorted Lists/21. Merge Two Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {

        // 1. 优先处理特殊情况，即l1和l2中存在一个链表为空的情况
        if(l1==NULL)
        return l2;
        if(l2==NULL)
        return l1;

        // 2. 使用双指针进行和归并排序中类似的归并操作
        ListNode *pos1=l1;      //链表1游标指针
        ListNode *pos2=l2;      //链表2游标指针
        ListNode *res=new ListNode(0);    //结果链表，同样使用了哑结点
        ListNode *respos=res;             //结果链表指针

        while(l1!=NULL&&l2!=NULL)
        {
            respos->next=new ListNode((l1->val<l2->val)?l1->val:l2->val);

            if(l1->val<l2->val)
            l1=l1->next;
            else
            l2=l2->next;
            respos=respos->next;
        }

        // 3. 处理两个链表中的剩下部分
        while(l1!=NULL)
        {
            respos->next=new ListNode(l1->val);

            l1=l1->next;
            respos=respos->next;
        }

        while(l2!=NULL)
        {
            respos->next=new ListNode(l2->val);

            l2=l2->next;
            respos=respos->next;
        }

        // 3. 删除结果链表的哑结点，返回最终结果
        ListNode *temp=res;
        res=res->next;
        delete temp;

        return res;
    }
};

File: Linked List/21. Merge Two Sorted Lists/21. Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        # 1. 处理两个链表中存在空链表的情况，若一个链表为空，则直接返回另一个链表
        if not l1:
            return l2
        if not l2:
            return l1
        
        # 2. 处理一般情况，创建结果数组，并创建三个指针——链表1的指针，链表2的指针以及结果链表的指针
        pos1=l1             # 链表1的指针
        pos2=l2             # 链表2的指针
        res=ListNode(0)     # 结果链表，包含哑结点
        respos=res          # 链表指针

        # 3. 首先，当两个指针均不指向空结点时，对两个指针指向的结点值进行比较，将较小值放入结果链表，然后更新加入结果的链表结点的指针以及结果链表的指针
        while pos1 and pos2:
            min=pos1.val if pos1.val<pos2.val else pos2.val
            respos.next=ListNode(min)
            
            if pos1.val<pos2.val:
                pos1=pos1.next
            else:
                pos2=pos2.next
            respos=respos.next

        # 3. 然后，若存在指针指向空结点，说明其中一个链表中的结点值已经全部加入结果链表中，这时将剩下的没有完全加入的链表中的结点依次加入结果链表中
        while pos1:
            respos.next=ListNode(pos1.val)

            pos1=pos1.next
            respos=respos.next
        
        while pos2:
            respos.next=ListNode(pos2.val)

            pos2=pos2.next
            respos=respos.next
        
        return res.next